rasa/core/policies/rule_policy.py

```diff
@@ -214,85 +310,125 @@ def predict_action_probabilities(
             and tracker.latest_action_name == active_form_name
         )
 
-        # A form has priority over any other rule.
-        # The rules or any other prediction will be applied only if a form was rejected.
-        # If we are in a form, and the form didn't run previously or rejected, we can
-        # simply force predict the form.
         if should_predict_form:
             logger.debug(f"Predicted form '{active_form_name}'.")
-            result[domain.index_for_action(active_form_name)] = 1
-            return result
+            return active_form_name
 
         # predict `action_listen` if form action was run successfully
         if should_predict_listen:
             logger.debug(
                 f"Predicted '{ACTION_LISTEN_NAME}' after form '{active_form_name}'."
             )
-            result[domain.index_for_action(ACTION_LISTEN_NAME)] = 1
-            return result
-
-        possible_keys = set(self.lookup.keys())
+            return ACTION_LISTEN_NAME
 
+    def _predict_rule_action(
+        self, tracker: DialogueStateTracker, domain: Domain
+    ) -> Optional[Text]:
         tracker_as_states = self.featurizer.prediction_states([tracker], domain)
         states = tracker_as_states[0]
 
         logger.debug(f"Current tracker state: {states}")
 
+        possible_keys = set(self.lookup[RULES].keys())
+        negative_keys = set(self.lookup[NEGATIVE_RULES].keys())
```

Comment:
should we only collect the negative keys if a form is active? Might be a good idea performance wise.